src/services/matcher.py

```diff
@@ -62,21 +62,41 @@ def create_distances_df(self,
         return closest_cats
 
     def _get_embeddings(self, entities: List[Entity]):
-        all_embeddings = [c.embeddings for c in entities]
+        all_embeddings = []
+        embeddings_belonging = []
+
+        for entity_id, entity in enumerate(entities):
+            for embedding in entity.embeddings:
+                all_embeddings.append(embedding)
+                embeddings_belonging.append(entity)
+
         all_embeddings = np.float32(np.vstack(all_embeddings))
         all_embeddings = normalize(all_embeddings, axis=1, norm="l2")
-        return all_embeddings
+        return all_embeddings, embeddings_belonging
 
     def filter_by_thr(self, closest: List[ClosestCats], thr: float):
         for cl in closest:
             filtered = [(c, d) for c, d in zip(cl.cats, cl.distances) if d < thr]
             cl.cats = [f for f, _ in filtered]
             cl.distances = [f for _, f in filtered]
             yield cl
+
+    @staticmethod
+    def reduce_closest_cats(closest_cats):
+        """deletes duplicates"""
+        reduced_closest_cats = {}
+        for cat in closest_cats:
+            if id(cat) not in reduced_closest_cats:
+                reduced_closest_cats[id(cat)] = cat
+        return reduced_closest_cats.values()
+
     def find_n_closest(self, for_check: List[Entity], stored_cats: List[Cat], max_n: int = 5, thr: float = 1):
-        emb_for_check = self._get_embeddings(for_check)
-        stored_emb = self._get_embeddings(stored_cats)
+        emb_for_check, _ = self._get_embeddings(for_check)
+        stored_emb, stored_emb_belonging  = self._get_embeddings(stored_cats)
         D, I = self.create_and_search_index(stored_emb[0].size, stored_emb, emb_for_check, k=max_n)
-        closest = self.create_distances_df(for_check, stored_cats, D, I)
-        res = list(self.filter_by_thr(closest, thr))
+        persons_and_matched_cats = self.create_distances_df(for_check, stored_emb_belonging, D, I)
+        for person_id, person in enumerate(persons_and_matched_cats):
+            person.cats = self.reduce_closest_cats(person.cats)
+            persons_and_matched_cats[person_id] = person
+        res = list(self.filter_by_thr(persons_and_matched_cats, thr))
         return res
```

Comment:
please use another logic
like postprocessing  as we discussed

Reply:
I don't understand what you mean by postprocessing.  I thought this was postprocessing

Reply:
discussed in chat